src/freezethaw.rs

```diff
@@ -0,0 +1,58 @@
+use linux_raw_sys::ioctl::FIFREEZE;
+use linux_raw_sys::ioctl::FITHAW;
+
+use rustix::fd::AsFd;
+use rustix::{io, ioctl};
+
+fn ioctl_fifreeze<Fd: AsFd>(fd: Fd) -> io::Result<()> {
+    // SAFETY: `FIFREEZE` is a no-argument opcode.
+    unsafe {
+        let ctl = ioctl::NoArg::<{ FIFREEZE }>::new();
+        ioctl::ioctl(fd, ctl)
+    }
+}
+
+fn ioctl_fithaw<Fd: AsFd>(fd: Fd) -> io::Result<()> {
+    // SAFETY: `FITHAW` is a no-argument opcode.
+    unsafe {
+        let ctl = ioctl::NoArg::<{ FITHAW }>::new();
+        ioctl::ioctl(fd, ctl)
+    }
+}
+
+/// syncfs() doesn't flush the journal on XFS,
+/// and since GRUB2 can't read the XFS journal, the system
+/// could fail to boot.
+///
+/// http://marc.info/?l=linux-fsdevel&m=149520244919284&w=2
+/// https://github.com/ostreedev/ostree/pull/1049
+///
+/// When calling this function you should be protected by SignalTerminationGuard
+/// to avoid being killed between `FIFREEZE` and `FITHAW`
+///
+/// This function always call syncfs() first, and
+/// calls `FIFREEZE` and `FITHAW` only for XFS filesystems
+pub(crate) fn fsfreeze_thaw_cycle<Fd: AsFd>(fd: Fd) -> io::Result<()> {
+    rustix::fs::syncfs(&fd)?;
+    let statfs = rustix::fs::fstatfs(&fd)?;
+    if statfs.f_type == libc::XFS_SUPER_MAGIC {
+        ioctl_fifreeze(&fd)?;
+        ioctl_fithaw(fd)?;
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use rustix::fs::{Mode, OFlags};
+    use rustix::io;
+
+    #[test]
+    #[ignore]
+    fn test_fsfreeze_thaw_cycle() -> io::Result<()> {
+        let dir = rustix::fs::open("/boot/", OFlags::RDONLY, Mode::RUSR)?;
```

Comment:
We shouldn't affect global state from unit tests. In the general case we should expect unit tests to run without privileges so this should always fail...

I'd just drop the unit test here.

(What I do want to do in general is have a standard pattern to add these types of things as integration tests, but it's a lot more complex)

Reply:
it has `#[ignore]`, so you can quickly play with the function, but I'm ok dropping it
For the record I confirmed with strace that it sent the right ioctl, and we can use subfolders, we don't need to use the mount point